tools: Run node_tests with coverage in parallel through node.js.

YashRE42 opened the pull request:
This PR is intended to supersede #20150.

Previously, we'd force node tests to run in serial mode. One might
think this is because it's not possible to consolidate coverage from
multiple runs, but nyc actually does this by default!

The default behaviour to merge multiple runs depended on using the
`--no-clean` argument, but given that there is no explicit guarantee
given in nyc documentation that `--no-clean` is safe to run in
parallel (ie would not suffer from eg concurrent write errors), and
realizing that the alternative (generating coverage from all parallel
runs in a separate folder and use `nyc merge` to combine them into a
single `coverage-final.json`) would be expensive to maintain, and also
be unable to leverage any of nyc's builtin optimizations for
instrumentation with parallel test runners (if any).
(The fact that nyc can handle child_processes is one of its main
differences from istanbul.js.)

Hence, in this commit, we reimplement parallelization for node tests
through node (this was previously through python) and make changes
that enable us to run node tests with coverage, in parallel.

An unintended benefit of this change is that stopping node tests via
keyboard interrupt (ctrl-C) now properly exits and ends all parallel
runners, where as previously they would continue to run (and print to
the terminal). This was only a minor nuisance as the tests would run
fast and end quickly anyway.

The net difference from this commit as measured by bash `time` when
node tests are run with coverage is:
real: ~ -14.486s
user: ~ +55.337s
sys: ~ +1.917s
(Lower values are better).

The net difference from this commit as measured by bash `time` when
node tests are run without coverage is:
real: ~ -0.7s
user: ~ -1.239s
sys: ~ +0.096s
(Lower values are better).

**Testing plan:** <!-- How have you tested? -->
I have verified that if say I change:
```diff
diff --git a/frontend_tests/node_tests/activity.js b/frontend_tests/node_tests/activity.js
index 6c2b5b04de..9cdcf26884 100644
--- a/frontend_tests/node_tests/activity.js
+++ b/frontend_tests/node_tests/activity.js
@@ -208,7 +208,7 @@ test("presence_list_full_update", ({override, mock_template}) => {
     override(padded_widget, "update_padding", () => {});
 
     mock_template("user_presence_rows.hbs", false, (data) => {
-        assert.equal(data.users.length, 7);
+        assert.equal(data.users.length, 200);
         assert.equal(data.users[0].user_id, me.user_id);
     });
```

Tests fail with the following output:
```bash
(zulip-py3-venv) re@DESKTOP-4C9KNPD:~/zulip$ ./tools/test-js-with-node --coverage
Starting node tests...
running test example2
running test stream_data
running test activity
running test navbar_alerts
running test notifications
running test stream_edit
running test example3
running test password
running test peer_data
running test example4
--------------------------------------------------
test failed: /home/re/zulip/frontend_tests/node_tests/activity.js > presence_list_full_update

AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:

7 !== 200

    at /home/re/zulip/frontend_tests/node_tests/activity.js:10:910
    at render (/home/re/zulip/frontend_tests/zjsunit/namespace.js:42:237)
    at BuddyList.items_to_html (/home/re/zulip/static/js/buddy_list.js:2:12205)
    at BuddyList.render_more (/home/re/zulip/static/js/buddy_list.js:9:696)
    at BuddyList.<anonymous> (/home/re/zulip/frontend_tests/node_tests/activity.js:6:119)
    at BuddyList.new_f [as fill_screen_with_content] (/home/re/zulip/frontend_tests/zjsunit/namespace.js:95:773)
    at BuddyList.populate (/home/re/zulip/static/js/buddy_list.js:9:76)
    at /home/re/zulip/static/js/activity.js:19:1461
    at Object.lib.measure_time (/home/re/zulip/frontend_tests/zjsunit/zblueslip.js:12:1088)
    at Object.build_user_sidebar (/home/re/zulip/static/js/activity.js:19:1343)
running test people

=============================== Coverage summary ===============================
Statements   : 25.98% ( 908/3494 )
Branches     : 15.62% ( 198/1267 )
Functions    : 22.82% ( 202/885 )
Lines        : 26.93% ( 888/3297 )
================================================================================

** Tests failed, PLEASE FIX! **
```

additionally, the coverage as reported by CI on this commit is:
https://github.com/zulip/zulip/runs/4241458795?check_suite_focus=true
![](https://user-images.githubusercontent.com/33805964/142288788-a9b6d002-e46d-4671-8566-c7a1af2f16d0.png)
which **is not the same** as the the coverage on main:
https://github.com/zulip/zulip/runs/4232577468?check_suite_focus=true
![](https://user-images.githubusercontent.com/33805964/142249415-669e6d23-1c3a-4869-9e12-4037abd11752.png)

But the difference is small and may just be due to adding `parallel_runner.js`?

**Profiling data** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
Values of `real` time as received from time(./tools/test-js-with-node --coverage)
| Run | first | second | third |
| ---- | ---- | ---- | ---- |
| main | 0m40.388s | 0m39.819s | 0m39.565s |
| python | 0m29.334s | 0m27.078s | 0m26.452s |
| node | 0m25.381s | 0m26.159s | 0m24.773s |

Values of `real` time as received from time(./tools/test-js-with-node)
run | first | second | third
-- | -- | -- | --
python (main) | 6.297s | 6.571s | 6.623s
node | 5.916s | 5.83s | 5.644s

Here's an .md with the data collected: [data.md](https://github.com/zulip/zulip/files/7556624/data.md)

YashRE42 commented on tools/test-js-with-node:
```diff
@@ -20,7 +20,7 @@ sanity_check.check_venv(__file__)
 from zulint.printer import BOLDRED, CYAN, ENDC, GREEN
 
-INDEX_JS = os.path.join(ROOT_DIR, "frontend_tests/zjsunit/index.js")
+PARALLEL_RUNNER_JS = os.path.join(ROOT_DIR, "frontend_tests/zjsunit/parallel_runner.js")
```
We can change this if we come up with a better name than "parallel_runner".

YashRE42 commented on frontend_tests/zjsunit/parallel_runner.js:
```diff
@@ -0,0 +1,54 @@
+const worker_processes = [];
+for (let i = 0; i < parallel; i += 1) {
+    const worker_process = child_process.spawn("node", [index, ...file_groups[i]]);
+    worker_process.on("error", (error) => {
+        // if we fail to create a child processes,
+        // exit and send a non zero return code, so we don't check coverage
+        console.error(error);
+        process.exit(2);
+    });
```
According to node [docs for the error event](https://nodejs.org/api/child_process.html#event-error):
"The 'exit' event may or may not fire after an error has occurred."

Since what we're doing in the handlers is `process.exit()` should we be concerned?

timabbott replied:
I don't think that's an issue -- the warning there is about doing the same thing on the `exit` and `error` events, which we're not doing.

andersk commented on frontend_tests/zjsunit/parallel_runner.js:
```diff
@@ -0,0 +1,54 @@
+        console.error(error);
+        process.exit(2);
+    });
+    worker_process.stdout.on("data", (data) => {
+        // use process.stdout.write to avoid adding extra new lines to logs
+        process.stdout.write(String(data));
+    });
+    worker_process.stderr.on("data", (data) => {
+        // use process.stderr.write to avoid adding extra new lines to logs
+        process.stderr.write("error!\n" + String(data));
+    });
```
Don’t manually shuffle data from pipes to stdout/stderr; use `spawn`/`fork` with `{stdio: "inherit"}` to avoid having pipes in the first place.

andersk commented on frontend_tests/zjsunit/parallel_runner.js:
```diff
@@ -0,0 +1,54 @@
+const file_groups = create_file_groups(files, parallel);
+const worker_processes = [];
+for (let i = 0; i < parallel; i += 1) {
+    const worker_process = child_process.spawn("node", [index, ...file_groups[i]]);
```
[`child_process.fork`](https://nodejs.org/api/child_process.html#child_processforkmodulepath-args-options) is cleaner for this case.

andersk commented on frontend_tests/zjsunit/parallel_runner.js:
```diff
@@ -0,0 +1,54 @@
+
+const file_groups = create_file_groups(files, parallel);
+const worker_processes = [];
+for (let i = 0; i < parallel; i += 1) {
```
`for (const file_group of file_groups)`

andersk commented on frontend_tests/zjsunit/index.js:
```diff
@@ -40,7 +40,17 @@ function immediate(f) {
+    // The only other case where this is possible is where you've
+    // removed all or almost all test files from /node_tests/
+    console.info(
+        "one or more child processes did not receive any files, one possibility is:\n" +
+            "Number of processes > number of test files.",
```
Or, we could not have that bug—just avoid creating empty groups in `create_file_groups`.

andersk commented on frontend_tests/zjsunit/parallel_runner.js:
```diff
@@ -0,0 +1,54 @@
+    throw new Error("No tests found");
+}
+
+const index = path.join(__dirname, "index.js");
```
`require.resolve("./index")`

andersk commented on frontend_tests/zjsunit/parallel_runner.js:
```diff
@@ -0,0 +1,54 @@
+            // or due to an error, etc
+            // exit this process (and all its children) and,
+            // send a non zero return code, so we don't check coverage
+            process.exit(1);
```
Forcing an `exit()` will leak all the remaining child processes to continue run in the background after we return to the user’s shell. Just set `process.exitCode = 1` instead.

andersk replied:
This comment still applies. Sending a termination signal to child processes is not the same as waiting for them to exit, so there could still be child processes running here. Use `process.exitCode`.

andersk commented on frontend_tests/zjsunit/parallel_runner.js:
```diff
@@ -0,0 +1,54 @@
+            process.exit(1);
+        }
+    });
+    worker_processes.push(worker_process);
```
This array is unused.

YashRE42 commented:
wip: this is currently broken.

YashRE42 committed:
tools: Run node_tests with coverage in parallel through node.js.

Previously, we'd force node tests to run in serial mode. One might
think this is because it's not possible to consolidate coverage from
multiple runs, but nyc actually does this by default!

The default behaviour to merge multiple runs depended on using the
`--no-clean` argument, but given that there is no explicit guarantee
given in nyc documentation that `--no-clean` is safe to run in
parallel (ie would not suffer from eg concurrent write errors), and
realizing that the alternative (generating coverage from all parallel
runs in a separate folder and use `nyc merge` to combine them into a
single `coverage-final.json`) would be expensive to maintain, and also
be unable to leverage any of nyc's builtin optimizations for
instrumentation with parallel test runners (if any).
(The fact that nyc can handle child_processes is one of its main
differences from istanbul.js.)

Hence, in this commit, we reimplement parallelization for node tests
through node (this was previously through python) and make changes
that enable us to run node tests with coverage, in parallel.

The net difference from this commit as measured by bash `time` when
node tests are run with coverage is:
real: ~ -14.486s
user: ~ +55.337s
sys: ~ +1.917s
(Lower values are better).

The net difference from this commit as measured by bash `time` when
node tests are run without coverage is:
real: ~ -0.7s
user: ~ -1.239s
sys: ~ +0.096s
(Lower values are better).

YashRE42 commented:
I had to drop all the changes that tried to improve the test failed or SIGINT behavior, so those are ~ still as bad as main.
Both before and after this PR:
- If a test fails, the other tests will continue to run and print to console.
- If the users sends SIGINT via `ctrl` + `c`, the tests won't end, and will instead continue to print to console.

I'd consider these cases to be just minor nuisances since the tests run fast and end pretty quickly anyway, but I'd have liked to solve the problem. My approach was:
```diff
diff --git a/frontend_tests/zjsunit/parallel_runner.js b/frontend_tests/zjsunit/parallel_runner.js
index b8441a524a..66ea89fe76 100644
--- a/frontend_tests/zjsunit/parallel_runner.js
+++ b/frontend_tests/zjsunit/parallel_runner.js
@@ -27,20 +27,97 @@ function create_file_groups(files, parallel) {
 
 const file_groups = create_file_groups(files, parallel);
 
+const worker_processes = [];
+
+// we use this map to ensure that we only call .kill() on the processes
+// we have forked, this is important because if we didn't ensure this,
+// it is entirely possible that we would send a kill signal to an
+// arbitrary process which just happens to have the same pid as the one
+// we had created (through pid reallocation).
+const has_exited = new Map();
+function kill_all_child_processes() {
+    for (const worker_process of worker_processes) {
+        const worker_process_pid = worker_process.pid;
+        if (!has_exited.get(worker_process_pid)) {
+            has_exited.set(worker_process_pid, true);
+            worker_process.kill();
+        }
+    }
+}
 for (const file_group of file_groups) {
     const worker_process = child_process.fork(index, file_group, {stdio: "inherit"});
+    const worker_process = child_process.spawn("node", [index, ...file_group], {stdio: "inherit"});
+    const current_pid = worker_process.pid;
+    console.log(current_pid + " has been spawned");
+    has_exited.set(current_pid, false);
     worker_process.on("error", (error) => {
         console.error(error);
-        // if we fail to create a child processes, send a non zero
+        // if we fail to create a child processes, kill all previously
+        // created processes, exit this process and send a non zero
         // return code, so that we don't check coverage
-        process.exitCode = 2;
+        has_exited.set(current_pid, true);
+        kill_all_child_processes();
+        process.exit(2);
     });
     worker_process.on("exit", (code) => {
+        has_exited.set(current_pid, true);
         if (code === undefined || code !== 0) {
             // if any worker_process ends without running all its test
-            // files eg due to an error, send a non zero return code,
-            // so that we don't check coverage
-            process.exitCode = 1;
+            // files eg due to an error, then kill all worker_processes,
+            // exit this process and send a non zero return code, so that
+            // we don't check coverage
+            kill_all_child_processes();
+            process.exit(1);
         }
     });
+    worker_processes.push(worker_process);
 }
+ process.on("SIGINT", () => {
+     kill_all_child_processes();
+ });
 ```
 But that didn't work, and I haven't been able to figure out a solution that works, yet.
 
 So, in the interest of time I've let the existing behavior survive and just implemented parallelization through node.

YashRE42 committed:
work_in_progress: Handle failed test cases cleanly.

YashRE42 committed:
debug: Cause example2.js test to always fail.

andersk commented on frontend_tests/zjsunit/parallel_runner.js:
```diff
@@ -0,0 +1,74 @@
+// arbitrary process which just happens to have the same pid as the one
+// we had created (through pid reallocation).
+// see: https://nodejs.org/api/child_process.html#subprocesskillsignal
+const has_exited = new Map();
```
Not necessary. Node already keeps track of which processes have exited and avoids signaling them. It won’t be fooled by PID reallocation.

YashRE42 replied:
> Sending a signal to a child process that has already exited is not an error but may have unforeseen consequences. Specifically, if the process identifier (PID) has been reassigned to another process, the signal will be delivered to that process instead which can have unexpected results.

The above is what the doc I linked says, there are also one or two SO answers that pointed me to this possibility, do you believe the docs are outdated? If so, then do you have a source that you / I could use to ask them to update?
